GitHub Issue NOAA-EMC/GSI#126.  Add use of high resolution Raob data.

jderber-NOAA opened the pull request:
This change includes the use of high resolution raob data into the GSI.  The data is read directly from file uprair without going into the prepbufr file.  Note that because of the many levels >2000 for some profiles, the standard bufrlib could not be used and the dynamic allocation version is used.  The default values for bufrlib are updated in glbsoi.  After all types of data are read, the memory for the bufrlib is released using exitbufr.

Some values that are defined in the creation of the prepbufr file are not defined for this data, so default values are used.  The observation errors are specified, but not much work has been done on them since a later project is already planned to do this.  The new observations are given types 119 (for q,t, and ps) and 219 and 218 for uv.  The observations given type 218 are those that do not have a pressure for the wind obs and only have heights (like pibals).

When there are both observations from the prepbufr file and from the high resolution data for the same station, the prepbufr data is moved to a monitored status and only the high resolution data is used.  Of course if the high resolution data is monitored, the prepbufr data will be used.

Two additional input files are modified (errtable and global_convinfo.txt).  They are located in /scratch1/NCEPDEV/da/John.Derber/git/gsi on hera.

To use the high definition data, the input file uprair should be pointed at the appropriate data (see Chris Hill) and the lines similar to: 
   uprair         ps          null        ps                  0.0     0     0
   uprair         t           null        t                   0.0     0     0
   uprair         q           null        q                   0.0     0     0  and/or
   uprair         uv          null        uv                  0.0     0     0

should be added to the OBS_INPUT namelist.  

One main additional routine was added (hdraobmod.f90).  Modifications were made to setupt,setupw,setupps, setupq, read_obs.F90 and glbsoi.f90.  In addition, modifications were made to many of the read routines to eliminate warning messages from the dynamic allocation version of bufrlib when files are closed when not open.  There was no such warning message for the non-dynamic allocation version.

CatherineThomas-NOAA commented on regression/regression_var.sh:
```diff
@@ -131,7 +131,7 @@ case $machine in
 
    #  On Hera, there are no scrubbers to remove old contents from stmp* directories.
    #  After completion of regression tests, will remove the regression test subdirecories
-   export clean=".true."
```
Was this change intended to be included?  I'm always suspicious when I see changes to the regression drivers.

jderber-NOAA replied:
You are correct.  That change was not intended.  I will remove it.

CatherineThomas-NOAA commented on modulefiles/modulefile.ProdGSI.hera:
```diff
@@ -8,7 +8,7 @@
 ##}
 #module-whatis " NOAA-EMC/GSI whatis description"
 
-setenv CRTM_FIX /scratch2/NCEPDEV/nwprod/NCEPLIBS/fix/crtm_v2.3.0
```
@jderber-NOAA  Is this CRTM path change related to this work or an accidental inclusion?

jderber-NOAA replied:
An accidental inclusion

jderber-NOAA commented:
Made a few changes that fixed 2 issues that Cathy pointed out.
Fixed issue with observation time.
Fixed issue in setupq.f90

Added qsat(i,j,k)=max(qmin,qsat(i,j,k)) to genqsat.
Fixes issue with convergence in some cases.

CatherineThomas-NOAA commented on src/gsi/cplr_gfs_ensmod.f90:
```diff
@@ -822,6 +822,10 @@ subroutine parallel_read_nemsio_state_(en_full,m_cvars2d,m_cvars3d,nlon,nlat,nsi
    if (k3u==0.or.k3v==0.or.k3t==0.or.k3q==0.or.k3oz==0) &  
       write(6,'(" WARNING, problem with one of k3-")')
 
+   do k=1,nsig
```
@jderber-NOAA Is this change related?

jderber-NOAA replied:
This change is not related, but fixed a bug in this routine.  It should have been in the master, since it was made in an earlier commit, but was not there and caused the regression tests to fail.  I will work with Mike to resolve it.

CatherineThomas-NOAA replied:
@jderber-NOAA Thanks, good to know.  I'm fine with keeping it in this PR as long as it's here intentionally.

jderber-NOAA commented:
Thanks!  I added a presentation to my issue which has revised figures
including descent data.

On Thu, May 6, 2021 at 10:02 AM CatherineThomas-NOAA <
***@***.***> wrote:

> *@CatherineThomas-NOAA* commented on this pull request.
> ------------------------------
>
> In src/gsi/cplr_gfs_ensmod.f90
> <https://github.com/NOAA-EMC/GSI/pull/149#discussion_r627449329>:
>
> > @@ -822,6 +822,10 @@ subroutine parallel_read_nemsio_state_(en_full,m_cvars2d,m_cvars3d,nlon,nlat,nsi
>     if (k3u==0.or.k3v==0.or.k3t==0.or.k3q==0.or.k3oz==0) &
>        write(6,'(" WARNING, problem with one of k3-")')
>
> +   do k=1,nsig
>
> @jderber-NOAA <https://github.com/jderber-NOAA> Thanks, good to know. I'm
> fine with keeping it in this PR as long as it's here intentionally.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/GSI/pull/149#discussion_r627449329>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASD2M5SYTQ2MSG32EWVRYQDTMKOQJANCNFSM423OC7DQ>
> .
>

CatherineThomas-NOAA commented on src/gsi/setupq.f90:
```diff
@@ -317,6 +318,34 @@ subroutine setupq(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_diagsav
      muse(i)=nint(data(iuse,i)) <= jiter
   end do
+!  If HD raobs available move prepbufr version to monitor
+  nhdq=0
```
@jderber-NOAA Last question: why does nhdq get initialized to zero while nhdt and the others do not?

jderber-NOAA replied:
You are correct.  I had found this error a few days ago and am awaiting instructions from Mike how to change it so I do not mess things up.

jderber-NOAA replied:
Changes have been made to the code to address this  issue and the one before it.

emilyhcliu commented:
@jderber-NOAA I remembered you told us in a recent obs meeting that you have made substantial changes in the NetCDF IO.  I did not see changes in NetCDF IO in this review.  The changes in NetCDF IO you referred to had already been merged into develop.  Is this correct?  We are about to add precipitation components in NetCDF IO.  We would like to add on top of the changes you made.  

I went through the code changes you made in this review.  They look good to me.

MichaelLueken-NOAA reviewed (commented):
Hi John,

I have completed my review of your changes.  There are some missing precisions for real values in hdraobmod.f90, questions related to commented out lines of code (if these lines are no longer necessary, please remove them), and unused variables.  The regression tests were run and all passed (or ran through to completion for the debug tests).  Once these issues have been addressed, I will be able to submit this work to the review committe.

Thanks,

Mike

jderber-NOAA commented on src/gsi/hdraobmod.f90:
```diff
@@ -0,0 +1,1315 @@
+  real(r_kind),dimension(nsig-1):: dpres
+  real(r_kind),dimension(maxlevs)::plevs
+  real(r_kind),allocatable,dimension(:,:):: cdata_all,cdata_out
+  real(r_kind) :: zob,tref
```
Mike,

Sorry there were so many issues.  I will try to do better next time.

I made changes to address all of your issues (I hope).  You noted that twindin and ctwind were not used.  This indicated that I had not included a check to make sure it was inside the time window.  When I added that some obs were rejected.  This was because for some observation the reported seconds in the input bufr data is sometimes missing.  I put in a check for this and if missing use 0.  This changed the values in output a bit (because the time of the observation changed for the better).

I also made a change in format statements in statsconv.f90 and statslight.f90 because the number of observations was exceeding what the format statement allowed.  I6 to I7.

MichaelLueken-NOAA commented:
The due date for the review committee to provide feedback has past, so I will give final approval to these changes and merge them to the NOAA-EMC/GSI master.
